### ISSUE TITLE
2D Tiff Image

### ISSUE THREAD

**Issue by sulaimanvesal** [Radiomics/pyradiomics#298]
Dear all,

Thanks for your effort and making such a great repository. As a newbie, I wanted to know whether this package supports 2D images as well? I could not find any example in notebooks folder. I would appreciate if you provide a guide about how to use this package for extraction of features for a 2D image (Tiff Image of breast Cancer).

**Comment by JoostJM** [comment]
@sulaimanvesal, PyRadiomics expects 3D Grayscale input (2D slices are allowed, as it is a 3D image with 1 of the dimensions of size 1).
[This](https://groups.google.com/forum/#!topic/pyradiomics/QLdD_qEw3PY) topic is of a similar questions, with some pointers and notes, including code to convert 2D to 3D (use with caution though, missing important meta-information).

**Comment by sulaimanvesal** [comment]
@JoostJM, thank you for the details and the links, I could figure it out through the examples and it works pretty well after adding a new dimension to 2d image. However, I am getting some Nan values for few of parameters like skewness, kurtosis and entropy along the error. I am wondering what would cause these errors.

Here is the output of my simple code:
```

 lowBound = min(parameterValues) - (min(parameterValues) % binwidth)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

	Minimum: 0.0
	TotalEnergy: 226060278.0
	StandardDeviation: 73.6490120246
	Median: 134.0
	Mean: 148.944797264
	Uniformity: nan
	Skewness: nan
	90Percentile: 255.0
	Kurtosis: nan
	Energy: 226060278.0
	Entropy: nan
	Maximum: 255.0
	Variance: 5424.1769722
	10Percentile: 60.0
	RobustMeanAbsoluteDeviation: 57.7256302086
	Range: 255.0
	MeanAbsoluteDeviation: 61.7252919842
	RootMeanSquared: 166.158748203
	InterquartileRange: 159.0
```

**Comment by JoostJM** [comment]
@sulaimanvesal, as far as I can see, the issue is related to the pixel type. 
If you read your image using simpleitk (`im = SimpleITK.ReadImage('path/to/image.tif')`) and print it out (`print(im)`), somewhere at the bottom it should show your pixel type. I think it will be a vector (3 color channels). If this is the case, that is the cause of your error, as PyRadiomics only accepts grayscale images (pixel type is scaler; i.e. integer, float, not a vector). How do plan to handle multiple values per pixel?

**Comment by sulaimanvesal** [comment]
@JoostJM Hi Joost, Sorry for late reply, as I was in the hospital for sometimes. I have looked and I can confirm that the type of my images is grayscale, here a small code and output. 

`from radiomics import firstorder, glcm, imageoperations, shape, glrlm, glszm, getTestCase
import SimpleITK as sitk
import six
import sys, os
import numpy as np
import cv2

imagename = 'input/Train_img/0-02.tiff'
maskname = 'input/Train_mask_GT/0-02.tiff'

image = sitk.ReadImage(imagename)
image3d = sitk.JoinSeries(image)
mask = sitk.ReadImage(maskname)
mask3d = sitk.JoinSeries(mask)

print(mask3d)

firstOrderFeatures = firstorder.RadiomicsFirstOrder(image3d,mask3d, label=255)
firstOrderFeatures.enableAllFeatures()  # On the feature class level, all features are disabled by default.
firstOrderFeatures.calculateFeatures()
for (key,val) in six.iteritems(firstOrderFeatures.featureValues):
  print("\t%s: %s" % (key, val))`

`Image (00000163E129FFB0)
  RTTI typeinfo:   class itk::Image<unsigned char,3>
  Reference Count: 1
  Modified Time: 1139
  Debug: Off
  Object Name: 
  Observers: 
    none
  Source: (none)
  Source output name: (none)
  Release Data: Off
  Data Released: False
  Global Release Data: Off
  PipelineMTime: 1129
  UpdateMTime: 1138
  RealTimeStamp: 0 seconds 
  LargestPossibleRegion: 
    Dimension: 3
    Index: [0, 0, 0]
    Size: [448, 448, 1]
  BufferedRegion: 
    Dimension: 3
    Index: [0, 0, 0]
    Size: [448, 448, 1]
  RequestedRegion: 
    Dimension: 3
    Index: [0, 0, 0]
    Size: [448, 448, 1]
  Spacing: [0.352778, 0.352778, 1]
  Origin: [0, 0, 0]
  Direction: 
1 0 0
0 1 0
0 0 1

  IndexToPointMatrix: 
0.352778 0 0
0 0.352778 0
0 0 1

  PointToIndexMatrix: 
2.83465 0 0
0 2.83465 0
0 0 1

  Inverse Direction: 
1 0 0
0 1 0
0 0 1

  PixelContainer: 
    ImportImageContainer (00000165E49D94D0)
      RTTI typeinfo:   class itk::ImportImageContainer<unsigned __int64,unsigned char>
      Reference Count: 1
      Modified Time: 1136
      Debug: Off
      Object Name: 
      Observers: 
        none
      Pointer: 00000165E4C29030
      Container manages memory: true
      Size: 200704
      Capacity: 200704

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 384, in getUniformityFeatureValue
    binEdges = imageoperations.getBinEdges(self.binWidth, self.targetVoxelArray)
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\imageoperations.py", line 38, in getBinEdges
    lowBound = min(parameterValues) - (min(parameterValues) % binwidth)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 350, in getKurtosisFeatureValue
    if m2 == 0:  # Flat Region
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 314, in getSkewnessFeatureValue
    if m2 == 0:  # Flat Region
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 121, in getEntropyFeatureValue
    binEdges = imageoperations.getBinEdges(self.binWidth, self.targetVoxelArray)
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\imageoperations.py", line 38, in getBinEdges
    lowBound = min(parameterValues) - (min(parameterValues) % binwidth)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

	Energy: 9427764.0
	RootMeanSquared: 143.161234734
	90Percentile: 255.0
	Mean: 115.943478261
	Range: 245.0
	Kurtosis: nan
	MeanAbsoluteDeviation: 69.5289981096
	Variance: 7052.24897921
	Minimum: 10.0
	RobustMeanAbsoluteDeviation: 71.3125844189
	Median: 88.0
	Skewness: nan
	StandardDeviation: 83.9776695271
	Maximum: 255.0
	Uniformity: nan
	TotalEnergy: 9427764.0
	Entropy: nan
	10Percentile: 35.0
	InterquartileRange: 98.75`

**Comment by JoostJM** [comment]
@sulaimanvesal Hmm strange... Are you able to provide me an anonymized image and mask to test?

**Comment by sulaimanvesal** [comment]
```
VectorImage (0000015EF1513610)
  RTTI typeinfo:   class itk::VectorImage<unsigned char,3>
  Reference Count: 1
  Modified Time: 954
  Debug: Off
  Object Name: 
  Observers: 
    none
  Source: (none)
  Source output name: (none)
  Release Data: Off
  Data Released: False
  Global Release Data: Off
  PipelineMTime: 944
  UpdateMTime: 953
  RealTimeStamp: 0 seconds 
  LargestPossibleRegion: 
    Dimension: 3
    Index: [0, 0, 0]
    Size: [448, 448, 1]
  BufferedRegion: 
    Dimension: 3
    Index: [0, 0, 0]
    Size: [448, 448, 1]
  RequestedRegion: 
    Dimension: 3
    Index: [0, 0, 0]
    Size: [448, 448, 1]
  Spacing: [1, 1, 1]
  Origin: [0, 0, 0]
  Direction: 
1 0 0
0 1 0
0 0 1

  IndexToPointMatrix: 
1 0 0
0 1 0
0 0 1

  PointToIndexMatrix: 
1 0 0
0 1 0
0 0 1

  Inverse Direction: 
1 0 0
0 1 0
0 0 1

  VectorLength: 4
  PixelContainer: 
    ImportImageContainer (00000160F2CFD1E0)
      RTTI typeinfo:   class itk::ImportImageContainer<unsigned __int64,unsigned char>
      Reference Count: 1
      Modified Time: 950
      Debug: Off
      Object Name: 
      Observers: 
        none
      Pointer: 00000160F2E0A2B0
      Container manages memory: true
      Size: 802816
      Capacity: 802816

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 121, in getEntropyFeatureValue
    binEdges = imageoperations.getBinEdges(self.binWidth, self.targetVoxelArray)
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\imageoperations.py", line 38, in getBinEdges
    lowBound = min(parameterValues) - (min(parameterValues) % binwidth)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 314, in getSkewnessFeatureValue
    if m2 == 0:  # Flat Region
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 384, in getUniformityFeatureValue
    binEdges = imageoperations.getBinEdges(self.binWidth, self.targetVoxelArray)
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\imageoperations.py", line 38, in getBinEdges
    lowBound = min(parameterValues) - (min(parameterValues) % binwidth)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

FAILED: Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\base.py", line 129, in calculateFeatures
    self.featureValues[feature] = getattr(self, 'get%sFeatureValue' % feature)()
  File "C:\Anaconda3\lib\site-packages\pyradiomics-0+unknown-py3.5-win-amd64.egg\radiomics\firstorder.py", line 350, in getKurtosisFeatureValue
    if m2 == 0:  # Flat Region
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

	Entropy: nan
	Kurtosis: nan
	Energy: 9427764.0
	Variance: 7052.24897921
	Mean: 115.943478261
	Minimum: 10.0
	StandardDeviation: 83.9776695271
	Maximum: 255.0
	RootMeanSquared: 143.161234734
	90Percentile: 255.0
	Median: 88.0
	RobustMeanAbsoluteDeviation: 71.3125844189
	Skewness: nan
	10Percentile: 35.0
	MeanAbsoluteDeviation: 69.5289981096
	Uniformity: nan
	TotalEnergy: 9427764.0
	InterquartileRange: 98.75
	Range: 245.0
```

**Comment by JoostJM** [comment]
@sulaimanvesal As I suspected, your input image is a `VectorImage` (`VectorLength: 4`), this is causing your error.

Before thinking about fixing this error, you first need to specify for yourself how you want to handle the fact that you have multiple values per pixel. The formulas in PyRadiomics assume 1 value per voxel...

**Comment by sulaimanvesal** [comment]
@JoostJM I could figure it out, you were right the image was VectorImage, what I just changed the image type like this:

```
image = sitk.ReadImage(imagename, sitk.sitkUInt8)
image3d = sitk.JoinSeries(image)
mask = sitk.ReadImage(maskname, sitk.sitkUInt8)
mask3d = sitk.JoinSeries(mask)
```

```
	Maximum: 116.0
	MeanAbsoluteDeviation: 24.4541398866
	Uniformity: 0.22449905482
	Entropy: 2.21501220989
	Median: 70.0
	RobustMeanAbsoluteDeviation: 19.581258488
	Range: 106.0
	Skewness: -0.232148663809
	InterquartileRange: 48.5
	TotalEnergy: 649963.0
	Mean: 69.5913043478
	RootMeanSquared: 75.1788013599
	10Percentile: 32.0
	Variance: 808.902533081
	Minimum: 10.0
	90Percentile: 107.0
	Kurtosis: 2.01498377921
	StandardDeviation: 28.4412118779
	Energy: 649963.0

```
This was the output of `first-order class` but for other classes such as GLCM and etc. the code exit without giving an error just this:

`Process finished with exit code -1073741819 (0xC0000005)`

Sorry for asking this silly questions but thank you for your fast responses. I will try to figure it out by myself.

**Comment by JoostJM** [comment]
@sulaimanvesal No problem! The fact that you don't get an error message is due to the fact that something went wrong in the C-extension (because it is C, it doesn't really have error handling). 

Do you have an example TIFF image and mask for me to debug with?

**Comment by sulaimanvesal** [comment]
@JoostJM I appreciated your help,

Here you can see a sample of TIFF image with mask and even DICOM slice.

[Sampl Image](https://www.dropbox.com/sh/5r0cobypa5k2b6u/AADgnNXAAQr59QYu5m01t896a?dl=0)

**Comment by JoostJM** [comment]
@sulaimanvesal, I see you have also your image available as DICOM. Is there a reason you try to extract from TIFF images? From DICOM you can also convert to NIFTII or NRRD, which is also grayscale. In fact, you can even provide a single slice DICOM to pyradiomics (DICOM is generally not supported because the entire volume has to be in 1 file).

**Comment by sulaimanvesal** [comment]
@JoostJM well, that is my problem, because I have the annotation only for one slice as ground truth. I made an algorithm to segment the lesions in an automatic way and I got good segmentation on 2D slices. My next aim was to extract some useful features from the segmented lesions to classify them. Yah well, I will go with DICOM, not TIFF. I hope it works with that :) Thanks alot for your follow up and inputs.

**Comment by JoostJM** [comment]
@sulaimanvesal, If you only have 1 slice ground truth (or at least a mask on one slice), inputting the DCM file directly is better (less steps changing the file / cleanest data).

In the meantime, I tested your images, but did not get the error you experienced. What version of PyRadiomics are you using? And what python version (python 2 or 3? 32 of 64 bits?)

What I did notice, is that the spacing, origin and direction of your dicom file are very different! This can affect the results (especially those in the `shape` class). I had to force your .tiff mask to the dcm file to extract features (one more reason for sticking to DCM in this case).

If at some point you want to extract volumetric feautures, I'd advise using the NIFTII or NRRD format, as these also contain the meta data for spacing and direction, and are grayscale by default.
